### ISSUE TITLE
security check on server startup

### ISSUE THREAD

**Issue by parkervcp** [pterodactyl/panel#895]
* Panel or Daemon: Daemon

Add server file checking.

Basically make a hash of the files/folders specified to verify that the server files are not modified. Take hash on install. This may be an issue with games after updates though. Will need a way to update the hash regularly.

This is to protect against people uploading mods or things that are not supported (game executable that mine coins)

**Comment by DaneEveritt** [comment]
If you can provide the necessary logic and solutions to the problems listed I can add this, but I do not see this being something I go out of my way to maintain or implement.

**Comment by parkervcp** [comment]
On install you hash the file/folder specified as a variable. Save that hash in the server.json.  
Before starting up recalc the hash on the folder.  
If the hash fails stop the startup process and error about changes to necessary game files.  
If the hash passes start the server regularly.

**Comment by Leopere** [comment]
Seems out of scope for a PHP application when you have software like [Tripwire](http://archive.is/uLnba).  

The other reason this seems kind of a bit moot is if you learn how to hack your way into this panel software the first thing you'd do is disable the startup checking feature.  If anything the process should be done by something more specialized to do this.

**Comment by DaneEveritt** [comment]
@Leopere I believe this is targeted more at making sure people aren't editing bash files used to start their server, not even for malicious reasons but to ensure that they don't end up breaking the boot process.

Processes are already isolated in containers. PHP wouldn't be doing any of this work, it would all be on the daemon side, but its also not a priority of mine right now.

**Comment by Leopere** [comment]
Yep, and that's what Tripwire does. I was suggesting that the best way to impose a Tripwire instance would be to have it running on a Read Only volume mentioning Docker since @parkervcp is working on a Docker implementation that it could be easily glued in beside that.  

Docker doesn't solve all execution issues it just isolates the issues into your web app which in this case is and would likely be the main focus of the machine its installed on.

You could add Tripwire and have it monitor the configs easily.  https://www.systutorials.com/docs/linux/man/4-twconfig/

**Comment by parkervcp** [comment]
@Leopere That would require the daemon interfacing with yet another thing and I am already sure we don't want to do that. Considering Tripwire costs money it's even more likely not going to be added. I know there are open source alternatives to these things but I was just looking for a quick to add thing that checks the file and not a full fledged file monitoring system.

**Comment by Leopere** [comment]
Just for the record Tripwire is free, however they do have a company that
sells stuff but there's a cli project that is open source. That we
discussed in discord.

**Comment by DaneEveritt** [comment]
Won't be addressing this issue.
